### PR TITLE
Avoid O(N^2) behavior in widenreturn

### PR DESCRIPTION
**Pull request by Keno**
In functions with lots of slots that construct PartialStructs with
lots of fields, `widenreturn` would attempt to repeatedly find an
interesting slot for Interconditional to use. This is unnecessary,
because the slot selected does not depend on the value of the
PartialStruct field. Memoize this computation and only do it once
per return statement.

### CODE REVIEW

**Comment by oscardssmith** [comment]
Is there a simple example where this is a performance win?

**Comment by Keno** [comment]
I've been working through all the pathologies of this testcase: https://gist.github.com/Keno/5587fbfe89bff96c863c8eeb1477defa. This was one of them.

File: base/compiler/abstractinterpretation.jl [hunk 1]
```diff
@@ -2075,7 +2086,7 @@ function widenreturn(@nospecialize(rt), @nospecialize(bestguess), nslots::Int, s
         local anyrefine = false
         for i in 1:length(fields)
             a = fields[i]
-            a = isvarargtype(a) ? a : widenreturn(a, bestguess, nslots, slottypes, changes)
+            a = isvarargtype(a) ? a : widenreturn(a, bestguess, nslots, slottypes, changes, interconditional_slot)
```
Comment (vtjnash) [inline base/compiler/abstractinterpretation.jl:2089]: Fields are prohibited from containing Conditional/InterConditional values

Comment (Keno) [reply base/compiler/abstractinterpretation.jl:2089]: Then the code that is here not is wrong, because it likes to add them.

**Comment by Keno** [comment]
Replaced by #44438

**Pull request closed without merging**